ppci/lang/c/semantics.py

```diff
@@ -692,21 +693,69 @@ def on_binop(self, lhs, op, rhs, location):
             result_typ = lhs.typ
             if not lhs.lvalue:
                 self.error("Expected lvalue", lhs.location)
+
+            if op[0] in ["+", "-"] and isinstance(lhs.typ, types.IndexableType):
+                self.ensure_integer(rhs.typ, location)
+                lhs.typ = self.ensure_no_void_ptr(lhs.typ, location)
             rhs = self.coerce(rhs, result_typ)
         elif op == ",":
             result_typ = rhs.typ
         elif op == "+":
-            # Handle pointer arithmatic
             if isinstance(lhs.typ, types.IndexableType):
-                rhs = self.coerce(rhs, self.int_type)
+                # indexable + integer
+                self.ensure_integer(rhs.typ, rhs.location)
+                lhs.typ = self.ensure_no_void_ptr(lhs.typ, lhs.location)
                 result_typ = lhs.typ
             elif isinstance(rhs.typ, types.IndexableType):
-                lhs = self.coerce(lhs, self.int_type)
+                # integer + indexable
+                self.ensure_integer(lhs.typ, lhs.location)
+                rhs.typ = self.ensure_no_void_ptr(rhs.typ, rhs.location)
                 result_typ = rhs.typ
+                # put indexable expression on the left to ease further processing
+                rhs, lhs = lhs, rhs 
+            else:
+                # numeric + numeric
+                result_typ = self.get_common_type(lhs.typ, rhs.typ, location)
+                lhs = self.coerce(lhs, result_typ)
+                rhs = self.coerce(rhs, result_typ)
+
+        elif op == "-":
+            if isinstance(lhs.typ, types.IndexableType):
+
+                if isinstance(rhs.typ, types.IndexableType):
+
+                    # indexable - indexable
+                    # operands must be of compatible types and should not be
+                    # 'void *'
+                    # type of result is a signed integer large enough to contain
+                    # pointer difference
+                    lhs.typ = self.ensure_no_void_ptr(lhs.typ, lhs.location)
+                    rhs.typ = self.ensure_no_void_ptr(rhs.typ, rhs.location)
+                    if not self._root_scope.equal_types(lhs.typ.element_type,
+                                                        rhs.typ.element_type):
+                        self.error(
+                            "operands must point to compatible types,"
+                            " got {} and {}".format(type_to_str(lhs.typ),
+                                                    type_to_str(rhs.typ)),
+                            location)
+                    # choose a signed integer able to hold a pointer difference
+                    if (self.context.sizeof(self.int_type) ==
+                        self.context.sizeof(lhs.typ)):
+                        result_typ = self.int_type
+                    else:
+                        result_typ = self.long_type
```

Comment:
This might be 4 bytes on LP64 mode?

Reply:
The purpose is to  find the C signed integer able to hold any difference between 2 pointers.
I assume that this is either int or long.

In i386, pointers and standard ints are 32bit,, "int" type is large enough.
In x86_64, pointers are 64bit and standard ints are 32bit, we need "long".

I think that finding the integer type able to contain a pointer value without loss will come again, and we could store this in the Csemantics context once for all.
These are the types size_t and ssize_t that we use at C level to index any array.